### PR TITLE
Fix some bugs in the references panel

### PR DESCRIPTION
**Pull request by chrismwendt**
I ran into some JavaScript errors when returning external refs from the new Go extension and I managed to change the grouping code enough to avoid these errors, but I admit I do not understand it very well and I may have introduced other problems. @sqs Can you take a look? :eyes:

### CODE REVIEW

File: shared/src/panel/views/HierarchicalLocationsView.tsx [hunk 1]
```diff
@@ -229,7 +225,7 @@ export class HierarchicalLocationsView extends React.PureComponent<Props, State>
                                                         </span>
                                                     </span>
                                                     <span className="badge badge-secondary badge-pill hierarchical-locations-view__item-badge">
-                                                        {locationsByGroup[i].get(groupName)!.length}
+                                                        {(locationsByGroup[i].get(groupName) || []).length}
```
Comment (chrismwendt) [inline shared/src/panel/views/HierarchicalLocationsView.tsx]: The `!` makes me think that `groupName` being a member of `locationsByGroup[i]` is an invariant that should not be violated. If that's true, then defaulting to `[]` here probably masks some other bug.

File: shared/src/panel/views/HierarchicalLocationsView.tsx [hunk 2]
```diff
@@ -182,10 +182,6 @@ export class HierarchicalLocationsView extends React.PureComponent<Props, State>
                     }
                     locs.push(loc)
                     locationsByGroup[i].set(group, locs)
-
-                    if (selectedGroups[i] !== group) {
-                        break
-                    }
```
Comment (chrismwendt) [inline shared/src/panel/views/HierarchicalLocationsView.tsx]: I couldn't figure out what purpose this serves, but `break`ing here prevents `orderedGroups[i]` for subsequent values of `i` from being set, which caused the `orderedGroups[i].includes(group)` call later on to throw an error. Is this some kind of optimization, or necessary for the semantics of grouping?

**Comment by sqs** [comment]
Will TAL soon

File: shared/src/panel/views/HierarchicalLocationsView.tsx (hunk 2 above)
Comment (sqs) [reply shared/src/panel/views/HierarchicalLocationsView.tsx]: This is for the following case.

URIs:

- `git://a#f1`
- `git://b#f2`

You want the component to show:

_ |  _ | _
---- | ----- | -----
**a**    | f1    | ...
b    |       |

You don't want to show `f2` when **a** is selected.

By deleting it, I think that `f2` will be shown in the list even if you've selected **a**.

File: shared/src/panel/views/HierarchicalLocationsView.tsx (hunk 1 above)
Comment (sqs) [reply shared/src/panel/views/HierarchicalLocationsView.tsx]: Yeah, I think it masks some other bug. Can you post the JS errors and the list of URIs that trigger the errors?

**Comment by sqs** [comment]
I will look into this

**Comment by chrismwendt** [comment]
I was running the master version of lang-go, which is not currently deployed to Sourcegraph.com.

**Prior to this PR being merged**

When visiting http://localhost:3080/github.com/gorilla/mux/-/blob/mux.go#L18:29&tab=references I got the error "Cannot read property 'length' of undefined":

image

https://github.com/sourcegraph/sourcegraph/blob/f624d9292538fa62ccf7c035a37722ea81bf5e8a/shared/src/panel/views/HierarchicalLocationsView.tsx#L232

After defaulting to `[]`, another error "Cannot read property 'includes' of undefined" occurred on the same link after clicking "Group by file" and selecting gorilla/mux with 0 results:

image

https://github.com/sourcegraph/sourcegraph/blob/f624d9292538fa62ccf7c035a37722ea81bf5e8a/shared/src/panel/views/HierarchicalLocationsView.tsx#L195

**After this PR was merged**

Now on master on the same link, I get the error "Cannot read property 'map' of undefined":

image

https://github.com/sourcegraph/sourcegraph/blob/f30fe606cb8b7d4703e7b7903eb1c1cb94f48ed1/shared/src/panel/views/HierarchicalLocationsView.tsx#L173

I'm going to take a look at this new bug.

**Pull request merged**